osquery/tables/system/windows/certificates.cpp

```diff
@@ -0,0 +1,300 @@
+            << " with " << GetLastError();
+    return;
+  }
+
+  dataBuff.resize(dataBuffLen, 0);
+  ret = CertGetCertificateContextProperty(
+      certContext, propId, dataBuff.data(), &dataBuffLen);
+
+  if (ret == 0) {
+    VLOG(1) << "Failed to get certificate property struct " << propId
+            << " with " << GetLastError();
+  }
+}
+
+// Enumerate and process a certificate store
+void enumerateCertStore(const HCERTSTORE& certStore,
+                        const std::string& certStoreName,
+                        std::set<std::string>& processedCerts,
+                        QueryData& results) {
+  auto certContext = CertEnumCertificatesInStore(certStore, nullptr);
+  if (certContext == nullptr) {
+    VLOG(1) << "Failed to enumerate certstore " << certStoreName << " with "
+            << GetLastError();
+    return;
+  }
+
+  while (certContext != nullptr) {
+    // Get the cert fingerprint, and ensure we haven't already processed it
+    std::vector<char> certHashbuff;
+    getCertCtxProp(certContext, CERT_HASH_PROP_ID, certHashbuff);
+    std::string fingerprint;
+    try {
+      boost::algorithm::hex(
+          std::string(certHashbuff.begin(), certHashbuff.end()),
+          back_inserter(fingerprint));
+    } catch (std::exception /* e */) {
+    }
+
+    if (processedCerts.find(fingerprint) != processedCerts.end()) {
+      certContext = CertEnumCertificatesInStore(certStore, certContext);
+      continue;
+    }
+    processedCerts.insert(fingerprint);
+
+    Row r;
+    r["sha1"] = fingerprint;
+    std::vector<char> certName(256, 0);
+    auto ret = CertGetNameString(certContext,
+                                 CERT_NAME_SIMPLE_DISPLAY_TYPE,
+                                 0,
+                                 nullptr,
+                                 certName.data(),
+                                 static_cast<unsigned long>(certName.size()));
+    r["common_name"] = certName.data();
+
+    auto subjSize = CertNameToStr(certContext->dwCertEncodingType,
+                                  &(certContext->pCertInfo->Subject),
+                                  CERT_SIMPLE_NAME_STR,
+                                  nullptr,
+                                  0);
+    std::vector<char> subject(subjSize, 0);
+    subjSize = CertNameToStr(certContext->dwCertEncodingType,
+                             &(certContext->pCertInfo->Subject),
+                             CERT_SIMPLE_NAME_STR,
+                             subject.data(),
+                             subjSize);
+    r["subject"] = subjSize == 0 ? "" : std::string(subject.data());
+
+    auto issuerSize = CertNameToStr(certContext->dwCertEncodingType,
+                                    &(certContext->pCertInfo->Issuer),
+                                    CERT_SIMPLE_NAME_STR,
+                                    nullptr,
+                                    0);
+    std::vector<char> issuer(issuerSize, 0);
```

Comment:
Instead of using several containers here, could you reuse one? The goal being to reduce the amount of memory a run of this method needs.